### PR TITLE
[RLlib] DreamerV3: docs

### PR DESCRIPTION
**Pull request by sven1977**
DreamerV3: docs
* Add DreamerV3 to algorithm table in docs (incl. some benchmark results)
* Add README for DreamerV3

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/dreamerv3/README.md [hunk 1]
```diff
@@ -1,27 +1,94 @@
...
-For more algorithm details, see:
+The implementation is based on RLlib's new Learner- and RLModule APIs and therefore allows
+for utilizing multi-GPU machines for training (see below for tips and tricks, example configs,
+and command lines).
+
+DreamerV3 trains a world model in supervised fashion using real environment
+interactions. The world model's objective is to correctly predict all aspects
+of the transition dynamics of the RL environment, which includes (besides predicting the
+correct next observations) predicting the received rewards as well as a boolean episode
+continuation flag.
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:15]:
```suggestion
of the transition dynamics of the RL environment, which includes predicting the
correct next observations, predicting the received rewards, and predicting a boolean for episode continuation.  
```

File: rllib/algorithms/dreamerv3/README.md [hunk 2]
```diff
@@ -1,27 +1,94 @@
...
-actual RL components of the model are never trained on actual environment data,
-but on dreamed trajectories only.
+## Overview
+An RLlib-based implementation for TensorFlow/Keras of the "DreamerV3" model-based reinforcement
+learning algorithm by D. Hafner et al. (Google DeepMind) 2023
 
-For more algorithm details, see:
+The implementation is based on RLlib's new Learner- and RLModule APIs and therefore allows
+for utilizing multi-GPU machines for training (see below for tips and tricks, example configs,
+and command lines).
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:9]:
```suggestion
This implementation allows scaling up the training by using multi-GPU machines for neural network updates (see below for tips and tricks, example configs,
and command lines).
```

File: rllib/algorithms/dreamerv3/README.md [hunk 3]
```diff
@@ -1,27 +1,94 @@
...
+
+DreamerV3 trains a world model in supervised fashion using real environment
+interactions. The world model's objective is to correctly predict all aspects
+of the transition dynamics of the RL environment, which includes (besides predicting the
+correct next observations) predicting the received rewards as well as a boolean episode
+continuation flag.  
+It utilizes a recurrent GRU-based architecture called "recurrent state space model" or RSSM.
+Alternatingly to training this world model, it is used to generate (or "dream") ficticious
+trajectories, with which two further models are trained in an RL-fashion:
+The critic and actor networks.
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:19]:
```suggestion

For more specific details about the algorithm refer to the original paper. 
```

File: rllib/algorithms/dreamerv3/README.md [hunk 4]
```diff
@@ -1,27 +1,94 @@
...
-interactions. The world model utilizes a recurrent GRU-based architecture
-("recurrent state space model" or RSSM) and uses it to predicts rewards,
-episode continuation flags, as well as, observations.
-With these predictions (dreams) made by the world model, both actor
-and critic are trained in classic REINFORCE fashion. In other words, the
-actual RL components of the model are never trained on actual environment data,
-but on dreamed trajectories only.
+## Overview
+An RLlib-based implementation for TensorFlow/Keras of the "DreamerV3" model-based reinforcement
+learning algorithm by D. Hafner et al. (Google DeepMind) 2023
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:5]: Link the arxiv paper

File: rllib/algorithms/dreamerv3/README.md [hunk 5]
```diff
@@ -1,27 +1,94 @@
...
+Just like in a standard policy gradient algorithm (e.g. REINFORCE), the critic tries to
+predict a correct value function (based on the world model-predicted rewards), whereas
+the actor tries to come up with good actions to take for maximizing accumulated rewards
+over time.
+In other words, the actual RL components of the model (actor and critic) are never
+trained on real environment data, but on dreamed trajectories only.
+
+## Note on Hyperparameter Tuning for DreamerV3
+DreamerV3 is an extremely versatile and stable algorithm that not only works well on
+different action- and observation spaces, including discrete and continuous actions, as well
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:29]:
```suggestion
different action and observation spaces (i.e. discrete and continuous actions, as well
```

File: rllib/algorithms/dreamerv3/README.md [hunk 6]
```diff
@@ -1,27 +1,94 @@
...
+as image and vector observations, but also almost needs no hyperparameter tuning, except for
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:30]:
```suggestion
as image and vector observations), but also has very little hyperparameters for  tuning.
```

File: rllib/algorithms/dreamerv3/README.md [hunk 7]
```diff
@@ -1,27 +1,94 @@
...
+a simple "model size" setting (from "XS" to "XL") and a value for the training ratio, which
+specifies how many steps to replay from the buffer for a training update vs how many
+steps to take in the actual environment.
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:33]:
```suggestion
All you need is a simple "model size" setting (from "XS" to "XL") and a value for the training ratio, which
specifies how many steps to replay from the buffer for a training update vs how many
steps to take in the actual environment.
```

File: rllib/algorithms/dreamerv3/README.md [hunk 8]
```diff
@@ -1,27 +1,94 @@
...
+
+
+## Note on multi-GPU Training with DreamerV3
+We found that when using multiple GPUs for DreamerV3 training, the following simple
+adjustments should be made to the standard config.
+- Multiply the original batch size (B=16) by the number of GPUs you are using.
+- Multiply the number of environments you sample from in parallel by the number of GPUs you are using.
+- Use a learning rate schedule for all learning rates (world model, actor, critic) with "priming".
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:41]: Can you link `priming` to the relevant arxiv link for the papers here?

File: rllib/algorithms/dreamerv3/README.md [hunk 9]
```diff
@@ -1,27 +1,94 @@
...
+Note that you can also switch on WandB logging with the above script via the options
+`--wandb-key=[your WandB API key] --wandb-project=[some project name] --wandb-run-name=[some run name]`
+
+
+## Results
+Our results on the Atari 100k and (visual) DeepMind Control Suite benchmarks match those
+reported in the paper.
+
+### Pong-v5 (100k) 1GPU vs 2GPUs vs 4GPUs
+<img src="https://github.com/ray-project/ray/tree/master/doc/source/rllib/images/dreamerv3/pong_1_2_and_4gpus.svg">
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/README.md:74]: This is not the best practice as this image will always be pointing to master. So if we release 2.7 and tomorrow change the image, the old content will point to the new image on master. 

We should do sth like: 
pong_1_2_and_4gpus

**Comment by kouroshHakha** [comment]
The docs are not getting build from some reason. :(

File: rllib/algorithms/dreamerv3/README.md [hunk 10]
```diff
@@ -1,27 +1,94 @@
...
+
+## Example Configs and Command Lines
+Use the config examples and templates in
+[the tuned_examples folder here](https://github.com/ray-project/ray/tree/master/rllib/tuned_examples/dreamerv3)
+in combination with the following scripts and command lines in order to run RLlib's DreamerV3 algorithm in your experiments:
+
+### Atari100k
+```shell
+$ cd ray/rllib/tests
+$ python run_regression_tests.py --dir ../tuned_examples/dreamerv3/atari_100k.py --env ALE/Pong-v5
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:56]: Yeah, good point. But we should add these WandB options to `rllib train` as well. These are super useful.

File: rllib/algorithms/dreamerv3/README.md (hunk 1 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:15]: fixed

File: rllib/algorithms/dreamerv3/README.md (hunk 2 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:9]: done

File: rllib/algorithms/dreamerv3/README.md (hunk 3 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:19]: fixed and added actual links to README.

File: rllib/algorithms/dreamerv3/README.md (hunk 4 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:5]: done

File: rllib/algorithms/dreamerv3/README.md (hunk 5 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:29]: done

File: rllib/algorithms/dreamerv3/README.md (hunk 7 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:33]: done

File: rllib/algorithms/dreamerv3/README.md (hunk 6 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:30]: done

File: rllib/algorithms/dreamerv3/README.md (hunk 9 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:74]: I think this is difficult doing this from the README (which is in a completely different location than the docs and does not get built together with the docs), but let me check.

File: rllib/algorithms/dreamerv3/README.md [hunk 11]
```diff
@@ -1,27 +1,94 @@
...
+as image and vector observations, but also almost needs no hyperparameter tuning, except for
+a simple "model size" setting (from "XS" to "XL") and a value for the training ratio, which
+specifies how many steps to replay from the buffer for a training update vs how many
+steps to take in the actual environment.
+
+
+## Note on multi-GPU Training with DreamerV3
+We found that when using multiple GPUs for DreamerV3 training, the following simple
+adjustments should be made to the standard config.
+- Multiply the original batch size (B=16) by the number of GPUs you are using.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md]: done

File: rllib/algorithms/dreamerv3/README.md (hunk 9 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md]: Figured it out. Never mind.

**Review by kouroshHakha — APPROVED** [review]

**Comment by kouroshHakha** [comment]
@sven1977 Can't merge yet: :( 

image

File: rllib/algorithms/dreamerv3/README.md (hunk 8 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/README.md:48]: done

**Pull request merged**